### ISSUE TITLE
Android standalone process can not  resolve host with c-ares

### ISSUE THREAD

**Issue by calvin2021y** [curl/curl#9647]
### I did this

use libcurl on android, for standalone process start by app.  

build with c-ares.

the verbose output:

```sh
Could not resolve host: wangwangim.oss-accelerate.aliyuncs.com
 Closing connection 2
```

after set CURLOPT_DNS_SERVERS to 1.1.1.1, it work without error.

### I expected the following

I expect it work with out call `CURLOPT_DNS_SERVERS`.   (I can not find a good dns server on a global geo location app)

### curl/libcurl version

curl-7_85_0

### operating system

Android.

**Comment by bagder** [comment]
This sounds like a c-ares problem, not a curl problem. It seems the c-ares version you use cannot find the default name resolver to use.

**Comment by bagder** [comment]
An experiment would be to see if the stock resolver works better.

**Comment by calvin2021y** [comment]
I am not test stock resolver yet.   c-ares use jvm to find the name server on android, but standalone process is not related to JVM. I guess that is why it not work.  

Is there backup way to find nameserver without JVM on android ?  

I want to avoid use stock resolver  because it block block on thread, I will try test it.

**Comment by calvin2021y** [comment]
with c-ares, If I set dns to `8.8.8.8,1.1.1.1`, when 8.8.8.8 blocked alway return resolve timeout, 1.1.1.1 is never used.

**Comment by calvin2021y** [comment]
with thread resolve it work without setup dns.

**Comment by calvin2021y** [comment]
I am test  "POSIX threaded resolver" and cares resolver,  compare with termux "ping".   same domain ping work and find a valid IP. but not for curl.

I am not sure why "POSIX threaded resolver" not work at the same time chrome and ping can work.   If I understood correct,  the "POSIX threaded resolver" use system call to get the IP ?

**Comment by calvin2021y** [comment]
"POSIX threaded resolver" work with WIFI.

with my 4G  return "Could not resolve host".   with 4G chrome work and "ping from termux" work.

**Comment by calvin2021y** [comment]
after use DOH `https://8.8.8.8/dns-query`, get this error:

```sh
connect to 8.8.8.8 port 443 failed: Connection refused
```

in the same time, curl is able to accept a http website with IP. (to avoid DNS).  

This cloud be some bug inside CURL because there is no DNS related.

**Comment by calvin2021y** [comment]
because in 4G mode https DOH with IP not work,  I try use a http DOH. (try a plain http DOH it work)

get this error:

```sh
Protocol "http" not supported or disabled in libcurl
 * Closing connection -1
 * Protocol "http" not supported or disabled in libcurl
 * Closing connection -1
```

**Comment by calvin2021y** [comment]
curl is able to connect to http with same IP at port 80,  not get timeout error with port 443. (in same case chrome is work)

I add this patch:
```c
      rc = connect(sockfd, &addr.sa_addr, addr.addrlen);
      fprintf( stderr, "curl connect %d, %d from %s:%d\n", rc, error, __FILE__, __LINE__);
```

```sh
* Hostname 8.8.8.8 was found in DNS cache
*   Trying 8.8.8.8:443...
curl connect -1, 0 from ~/src/curl/lib/connect.c:1334
* connect to 8.8.8.8 port 443 failed: Connection refused
* Failed to connect to 8.8.8.8 port 443 after 1004 ms: Connection refused
* Closing connection 3
```

**Comment by bagder** [comment]
Sorry, you seem to just blurt out all your tests and results here. I don't see how they help us take this forward.

curl has no "magic" tricks for name resolving, it either uses the libc functions or it uses c-ares for doing it. The similar situation goes for connecting to port 443 on a remote server. If it gets connection refused, that's what the TCP/IP stack says.

if you want to learn how the tools differ and why one works and another doesn't, then I suggest you Wireshark the connections and learn how exactly they differ over the wire.

**Comment by calvin2021y** [comment]
Thanks for the advise. the problem is it only not work for 4G, I am not sure how Wireshark is able to help this android 4G in device network.

**Comment by bagder** [comment]
> I am not sure how Wireshark is able to help this android 4G in device network.

I don't know what tools there are for Android to do the same stuff, I just assume they exist.

> cloud it be openssl related ? since http work

The connection is refused at TCP level, nothing even tried to use TLS yet. Something blocks the port most likely.

**Comment by timprepscius** [comment]
Was there any more information on this?

I am having a most difficult time getting curl function correctly with android

**Comment by bagder** [comment]
@timprepscius if you see the exact same thing, the problem is still presumed to be in c-ares.

Write a small example that resolves host names with c-ares and verify that it works. Then we know c-ares works and I'm wrong and the problem is in curl. I just think everything points to a problem within c-ares.

**Comment by timprepscius** [comment]
All of my curl queries were res 6.
It turned out that I was missing the:
```
<uses-permission android:name="android.permission.INTERNET" />
<uses-permission android:name="android.permission.ACCESS_NETWORK_STATE" />
```
in the android manifest.

Boy, 20 hours of trying every possible build option for dns resolution. different versions, different ssls.  That was exhausting.
Now I have to figure out what Vulkan capability is not actually supported, even though it claims it is.  I wish there were better error codes than "6" for curl and  "Incomplete" for Vulkan.

For future android dev.  If you see res 6 on android, first check your permissions. And Vulkan - your shaders might need to be rewritten to work on Android.  Blech.

**Comment by leleliu008** [comment]
I also encountered this when using the curl command-line on Android-12, I resolved it by passing `--dns-servers 1.1.1.1` to curl command, if no this option, it always report me `curl: (6) Could not resolve host: xxx`

I also built wget and linked libcares, it works without `--dns-servers=`
